File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public double[] solution(int k, int[][] ranges) {
        // 1. 우박수열(y값) 생성
        List<Integer> yList = new ArrayList<>();
        yList.add(k);
        while (k > 1) {
            if (k % 2 == 0) {
                k /= 2;
            } else {
                k = k * 3 + 1;
            }
            yList.add(k);
        }

        int n = yList.size() - 1; // 총 이동 횟수 (n)
        
        // 2. 각 구간의 정적분(넓이) 미리 계산
        double[] areas = new double[n];
        for (int i = 0; i < n; i++) {
            double y1 = yList.get(i);
            double y2 = yList.get(i + 1);
            areas[i] = (y1 + y2) / 2.0; // 사다리꼴 넓이 공식 (윗변+아랫변)/2 * 높이(1)
        }

        // 3. 누적합 배열 생성 (구간 합을 빠르게 구하기 위함)
        double[] prefixSum = new double[n + 1];
        prefixSum[0] = 0;
        for (int i = 0; i < n; i++) {
            prefixSum[i + 1] = prefixSum[i] + areas[i];
        }

        // 4. 각 범위(ranges)에 대한 정적분 결과 계산
        double[] answer = new double[ranges.length];
        for (int i = 0; i < ranges.length; i++) {
            int start = ranges[i][0];
            int end = n + ranges[i][1]; // 음수 범위 고려

            if (start > end) {
                answer[i] = -1.0; // 유효하지 않은 구간
            } else {
                answer[i] = prefixSum[end] - prefixSum[start];
            }
        }

        return answer;
    }
}

File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.js
function solution(k, ranges) {
    // 1. 우박수열(y값) 배열 생성
    const ySequence = [k];
    while (k > 1) {
        k = k % 2 === 0 ? k / 2 : k * 3 + 1;
        ySequence.push(k);
    }
    
    const n = ySequence.length - 1; // 총 이동 횟수
    const areas = [];
    
    // 2. 각 단위 구간별 사다리꼴 넓이 미리 계산
    for (let i = 0; i < n; i++) {
        const h1 = ySequence[i];
        const h2 = ySequence[i+1];
        areas.push((h1 + h2) / 2); // 윗변 h1, 아랫변 h2, 높이 1인 사다리꼴
    }
    
    // 3. 정적분 결과 계산
    return ranges.map(([start, end]) => {
        const actualEnd = end <= 0 ? n + end : end; // 0 이하의 끝점 변환
        
        if (start > actualEnd) return -1.0; // 시작점이 끝점보다 크면 유효하지 않은 구간
        if (start === actualEnd) return 0.0;  // 시작점과 끝점이 같으면 넓이 0
        
        let sum = 0;
        for (let i = start; i < actualEnd; i++) {
            sum += areas[i];
        }
        return sum;
    });
}
